File: plugins/datasources/yesoreyeram-supabase-datasource/src/filter.ts
import { SelectableValue } from '@grafana/data';

// ---------------------------------------------------------------------------
// Filter model
//
// The structured filter tree is compiled into PostgREST query parameters on the
// backend. The frontend only describes conditions/groups; it does not build the
// query parameters itself.
// ---------------------------------------------------------------------------

export type LogicalConnector = 'and' | 'or';

export interface FilterCondition {
  kind: 'condition';
  /** Column name. */
  field: string;
  /** Operator (see OPERATORS). */
  op: string;
  /** Comparison value. Ignored for unary operators. */
  value?: string;
}

export interface FilterGroup {
  kind: 'group';
  /** How the direct children are combined (AND/OR). */
  connector: LogicalConnector;
  children: FilterNode[];
}

export type FilterNode = FilterCondition | FilterGroup;

export function newCondition(): FilterCondition {
  return { kind: 'condition', field: '', op: 'eq', value: '' };
}

export function newGroup(): FilterGroup {
  return { kind: 'group', connector: 'and', children: [] };
}

export function emptyRootGroup(): FilterGroup {
  return newGroup();
}

// ---------------------------------------------------------------------------
// Operator catalog
// ---------------------------------------------------------------------------

export interface OperatorDef {
  value: string;
  label: string;
  /**
   * 'none'   = unary (no value)
   * 'single' = one value
   */
  arity: 'none' | 'single';
}

// Operator tokens map directly onto PostgREST operators. A `not.` prefix negates
// the operator; the `is` family carries its target inline (is.null / is.true /
// …). The backend (filter.go) compiles these into query parameters.
const EQUALITY: OperatorDef[] = [
  { value: 'eq', label: '=', arity: 'single' },
  { value: 'neq', label: '!=', arity: 'single' },
];

const COMPARISON: OperatorDef[] = [
  { value: 'gt', label: '>', arity: 'single' },
  { value: 'gte', label: '>=', arity: 'single' },
  { value: 'lt', label: '<', arity: 'single' },
  { value: 'lte', label: '<=', arity: 'single' },
];

const TEXT_MATCH: OperatorDef[] = [
  { value: 'like', label: 'like (use * or %)', arity: 'single' },
  { value: 'ilike', label: 'ilike (case-insensitive)', arity: 'single' },
  { value: 'match', label: 'matches regex (~)', arity: 'single' },
  { value: 'imatch', label: 'matches regex, case-insensitive (~*)', arity: 'single' },
];

const LIST: OperatorDef[] = [
  { value: 'in', label: 'in (comma-separated list)', arity: 'single' },
];

const ARRAY: OperatorDef[] = [
  { value: 'cs', label: 'contains (@>)', arity: 'single' },
  { value: 'cd', label: 'contained in (<@)', arity: 'single' },
];

const NULLABLE: OperatorDef[] = [
  { value: 'is.null', label: 'is null', arity: 'none' },
  { value: 'not.is.null', label: 'is not null', arity: 'none' },
  { value: 'is.true', label: 'is true', arity: 'none' },
  { value: 'is.false', label: 'is false', arity: 'none' },
];

const ALL_OPERATORS: OperatorDef[] = [
  ...EQUALITY,
  ...COMPARISON,
  ...TEXT_MATCH,
  ...LIST,
  ...ARRAY,
  ...NULLABLE,
];

export function operatorsForType(_type?: string): OperatorDef[] {
  return ALL_OPERATORS;
}

export function operatorOptions(type?: string): Array<SelectableValue<string>> {
  return operatorsForType(type).map((o) => ({ label: o.label, value: o.value }));
}

export function operatorArity(op: string): 'none' | 'single' {
  return ALL_OPERATORS.find((o) => o.value === op)?.arity ?? 'single';
}

// ---------------------------------------------------------------------------
// Template interpolation
// ---------------------------------------------------------------------------

export type ValueInterpolator = (value: string) => string;

export function interpolateFilterTree(root: FilterGroup, interpolate: ValueInterpolator): FilterGroup {
  const mapNode = (node: FilterNode): FilterNode => {
    if (node.kind === 'group') {
      return { ...node, children: node.children.map(mapNode) };
    }
    if (!node.value) {
      return node;
    }
    return { ...node, value: interpolate(node.value) };
  };
  return { ...root, children: root.children.map(mapNode) };
}

// ---------------------------------------------------------------------------
// JSON (de)serialization for persistence on the query
// ---------------------------------------------------------------------------

export function parseFilterTree(raw?: string): FilterGroup {
  if (!raw) {
    return emptyRootGroup();
  }
  try {
    const parsed = JSON.parse(raw);
    if (parsed && parsed.kind === 'group' && Array.isArray(parsed.children)) {
      return parsed as FilterGroup;
    }
  } catch {
    // ignore malformed persisted state
  }
  return emptyRootGroup();
}

export function stringifyFilterTree(root: FilterGroup): string {
  if (root.children.length === 0) {
    return '';
  }
  return JSON.stringify(root);
}

File: plugins/datasources/yesoreyeram-supabase-datasource/pkg/plugin/filter.go
package plugin

import (
	"strings"
)

// FilterNode is a node in the structured filter tree sent from the query editor.
// A node is either a condition (kind="condition") or a group (kind="group").
//
// Unlike a SQL where clause, PostgREST filters are URL query parameters. A
// top-level AND group compiles to one parameter per condition (`age=gt.18`,
// implicit AND); a top-level OR group compiles to a single `or=(...)` parameter;
// nested groups compile to inline `and(...)`/`or(...)` expressions inside their
// parent's parameter.
type FilterNode struct {
	Kind string `json:"kind"`

	// Condition fields.
	Field string `json:"field,omitempty"`
	// Op is the operator token. It may carry a `not.` negation prefix
	// (e.g. `not.eq`) and the `is` family carries its target inline
	// (`is.null`, `is.true`, `is.false`, `is.unknown`, `is.not_null`).
	Op    string `json:"op,omitempty"`
	Value string `json:"value,omitempty"`

	// Group fields.
	Connector string       `json:"connector,omitempty"`
	Children  []FilterNode `json:"children,omitempty"`
}

// SortItem is a single sort directive: a column name and a direction.
type SortItem struct {
	Field     string `json:"field"`
	Direction string `json:"direction"`
}

// Param is a URL query parameter key-value pair.
type Param struct {
	Key   string
	Value string
}

// BuildParams converts a structured filter tree (the root group) into PostgREST
// query parameters. It returns nil when there are no valid conditions, so the
// caller can omit the filter entirely.
//
// See https://postgrest.org/en/stable/references/api/tables_views.html#horizontal-filtering
func BuildParams(root FilterNode) []Param {
	if root.Kind != "group" {
		return nil
	}
	return compileGroupParams(&root)
}

// connectorOf returns the normalised logical connector ("and"/"or") for a group,
// defaulting to "and".
func connectorOf(group *FilterNode) string {
	if strings.EqualFold(group.Connector, "or") {
		return "or"
	}
	return "and"
}

// compileGroupParams compiles a group in a top-level (query-parameter) context.
//
//   - An OR group becomes a single `or=(...)` parameter whose body is the
//     comma-joined inline serialisation of its children.
//   - An AND group becomes one parameter per child: conditions map to
//     `field=op.value`; nested AND groups are flattened into the same context;
//     nested OR groups become their own `or=(...)` parameter.
func compileGroupParams(group *FilterNode) []Param {
	if connectorOf(group) == "or" {
		parts := serializeGroupParts(group)
		if len(parts) == 0 {
			return nil
		}
		return []Param{{Key: "or", Value: "(" + strings.Join(parts, ",") + ")"}}
	}

	params := make([]Param, 0, len(group.Children))
	for i := range group.Children {
		child := group.Children[i]
		if child.Kind == "group" {
			if connectorOf(&child) == "and" {
				// AND nested in AND: flatten into the same parameter list.
				params = append(params, compileGroupParams(&child)...)
				continue
			}
			parts := serializeGroupParts(&child)
			if len(parts) == 0 {
				continue
			}
			params = append(params, Param{Key: "or", Value: "(" + strings.Join(parts, ",") + ")"})
			continue
		}
		if field, body, ok := conditionBody(child, false); ok {
			params = append(params, Param{Key: field, Value: body})
		}
	}
	if len(params) == 0 {
		return nil
	}
	return params
}

// serializeGroupParts returns the inline serialisation of each valid child of a
// group (used when the group is nested inside another logical operator).
func serializeGroupParts(group *FilterNode) []string {
	parts := make([]string, 0, len(group.Children))
	for i := range group.Children {
		child := group.Children[i]
		var s string
		if child.Kind == "group" {
			s = serializeInlineGroup(&child)
		} else {
			s = serializeInlineCondition(child)
		}
		if s != "" {
			parts = append(parts, s)
		}
	}
	return parts
}

// serializeInlineGroup renders a nested group as an inline `and(...)`/`or(...)`
// expression. A group that reduces to a single condition drops the wrapper.
func serializeInlineGroup(group *FilterNode) string {
	parts := serializeGroupParts(group)
	if len(parts) == 0 {
		return ""
	}
	if len(parts) == 1 {
		return parts[0]
	}
	return connectorOf(group) + "(" + strings.Join(parts, ",") + ")"
}

// serializeInlineCondition renders a condition as `field.op.value`, the form used
// inside `or(...)`/`and(...)` groups. Values are quoted when they contain
// reserved characters so the surrounding group is not mis-parsed.
func serializeInlineCondition(c FilterNode) string {
	field, body, ok := conditionBody(c, true)
	if !ok {
		return ""
	}
	return field + "." + body
}

// conditionBody compiles a single condition to its (field, body) pair, where the
// body is the PostgREST operator expression after the column name (e.g.
// `eq.Alice`, `not.gt.18`, `is.null`, `in.(a,b)`). It returns ok=false when the
// condition is incomplete (no field, or a value-requiring operator with no
// value). When inGroup is true, values are quoted if they contain reserved
// characters.
func conditionBody(c FilterNode, inGroup bool) (string, string, bool) {
	field := strings.TrimSpace(c.Field)
	if field == "" {
		return "", "", false
	}

	op := strings.TrimSpace(c.Op)
	if op == "" {
		op = "eq"
	}
	value := strings.TrimSpace(c.Value)

	// Split an optional leading `not.` negation prefix.
	negated := false
	if strings.HasPrefix(op, "not.") {
		negated = true
		op = strings.TrimPrefix(op, "not.")
	}

	var body string
	switch {
	case op == "is" || strings.HasPrefix(op, "is."):
		// Unary IS check against null/not_null/true/false/unknown.
		target := "null"
		if strings.HasPrefix(op, "is.") {
			target = strings.TrimPrefix(op, "is.")
		} else if value != "" {
			target = strings.ToLower(value)
		}
		switch target {
		case "null", "not_null", "true", "false", "unknown":
		default:
			target = "null"
		}
		body = "is." + target
	case op == "in":
		list := buildInList(value)
		if list == "()" {
			return "", "", false
		}
		body = "in." + list
	default:
		// Binary operator (eq, neq, gt, gte, lt, lte, like, ilike, match,
		// imatch, cs, cd, fts, ...). All require a value.
		if value == "" {
			return "", "", false
		}
		body = op + "." + formatValue(value, inGroup)
	}

	if negated {
		body = "not." + body
	}
	return field, body, true
}

// formatValue renders a scalar value for an operator expression. Inside a
// logical group the value is quoted if it contains reserved characters so the
// group's commas/parentheses are not mis-parsed; at the top level the value is
// passed through unchanged (preserving like/ilike `*`/`%` patterns).
func formatValue(value string, inGroup bool) string {
	if inGroup {
		return quoteIfReserved(value)
	}
	return value
}

// buildInList renders the comma-separated value of an `in` operator into the
// PostgREST list form `(a,b,c)`. Each token is quoted when it contains reserved
// characters. Tokens that themselves need to contain a comma cannot be expressed
// through the comma-separated editor input.
func buildInList(value string) string {
	tokens := make([]string, 0)
	for _, t := range strings.Split(value, ",") {
		t = strings.TrimSpace(t)
		if t == "" {
			continue
		}
		tokens = append(tokens, quoteIfReserved(t))
	}
	return "(" + strings.Join(tokens, ",") + ")"
}

// quoteIfReserved wraps a value in double quotes when it contains a PostgREST
// reserved character (`,`, `(`, `)`) or already starts with a quote, escaping
// embedded double quotes. Plain values are returned unchanged.
func quoteIfReserved(value string) string {
	if value == "" {
		return value
	}
	if strings.ContainsAny(value, ",()") || strings.HasPrefix(value, `"`) {
		return `"` + strings.ReplaceAll(value, `"`, `\"`) + `"`
	}
	return value
}
